End to end support for bfp16 mac/mul intrinsics

niwinanto opened the pull request.

niwinanto committed:
[AIE2P] Support bfp16 mac/mul intrinsics - frontend

niwinanto committed:
[AIE2P] Support bfp16 mac/mul intrinsics - backend

andcarminati commented on llvm/lib/Target/AIE/aie2p/AIE2PInstructionSelector.cpp:
```diff
@@ -2910,6 +2920,88 @@ AIE2PInstructionSelector::getCombinedOpcodeSRSUPS(
+  case Intrinsic::aie2p_BFP576_BFP1152_ACC2048_mul_conf:
+  case Intrinsic::aie2p_BFP576_BFP1152_ACC2048_negmul_conf:
+    MI = MIB.buildInstr(TII.getOpCode(I), {DstReg},
+                        {Src1Reg, Src2Reg, /*Mode*/ I.getOperand(8).getReg()});
```
Clang-tidy can help with those comments, like: `/*Mode=*/I.getOperand(8).getReg()}`

andcarminati commented on llvm/lib/Target/AIE/aie2p/AIE2PInstructionSelector.cpp:
```diff
@@ -2910,6 +2920,88 @@ AIE2PInstructionSelector::getCombinedOpcodeSRSUPS(
+          .addImm(AIE2P::sub_bfp576_lo)
+          .addReg(Src2RegSub1)
+          .addImm(AIE2P::sub_bfp576_hi);
+  constrainOperandRegClass(*MF, TRI, MRI, TII, RBI, *RegSeq1,
```
nit: can we have a lambda where we pass something like `constrainRegSeq(RegSeq1, AIE2P::VEC512RegClass, 1);` ?

niwinanto replied:
Done!

andcarminati commented on llvm/lib/Target/AIE/aie2p/AIE2PInstructionSelector.cpp:
```diff
@@ -18,6 +18,8 @@
 #include "AIE2PTargetMachine.h"
 #include "AIEBaseInstructionSelector.h"
 #include "MCTargetDesc/aie2p/AIE2PMCTargetDesc.h"
+#include "llvm/ADT/ArrayRef.h"
```
Those includes are not necessary here.

khallouh reviewed (commented):
Could you add register bank select tests for these intrinsics? No need to test every single type, only each intrinsic would be enough, as all vector types follow the same code path.

niwinanto commented:
> Could you add register bank select tests for these intrinsics? No need to test every single type, only each intrinsic would be enough, as all vector types follow the same code path.

Done!

khallouh commented on clang/lib/Headers/aie2p_vmult.h:
```diff
@@ -38827,6 +38827,376 @@ addmsc_elem_64_conf(v64bfloat16 a, int sgn_x, v64bfloat16 b, int sgn_y,
                                                             conf);
 };
```
Nit: 2025 for this file?

khallouh commented on llvm/lib/Target/AIE/aie2p/AIE2PInstrPatterns.td:
```diff
@@ -301,6 +301,45 @@ def : PatInaccessibleMem<(int_aie2p_I512_I512_ACC512_bf_mac_conf VEC512:$s1, VEC
 def : PatInaccessibleMem<(int_aie2p_I512_I512_ACC512_bf_mul_conf VEC512:$s1, VEC512:$s2, eR:$acc),
            (EXTRACT_SUBREG (VMUL_f_vmul_bf_vmul_bf_core_X_X VEC512:$s1, VEC512:$s2, eR:$acc), sub_512_acc_lo)>;
```
Nit: 2024-2025 in the header

khallouh commented on llvm/lib/Target/AIE/aie2p/AIE2PInstructionSelector.cpp:
```diff
@@ -2973,6 +2984,88 @@ AIE2PInstructionSelector::getCombinedOpcodeSRSUPS(
+  ConstrainRegSeq(*RegSeq2, AIE2P::EXPVEC64RegClass, 3);
+  ConstrainRegSeq(*RegSeq3, AIE2P::VEC512RegClass, 1);
+  ConstrainRegSeq(*RegSeq3, AIE2P::EXPVEC64RegClass, 3);
+  ConstrainRegSeq(*RegSeq4, AIE2P::mEXaRegClass, 1);
```
Nit: VEC576 instead of mEXa. You could add this if not already there.

khallouh commented on llvm/lib/Target/AIE/aie2p/AIE2PRegisterBankInfo.cpp:
```diff
@@ -869,6 +869,9 @@ AIE2PRegisterBankInfo::getRegBankFromRegClass(const TargetRegisterClass &RC,
   case AIE2P::mS3RegClassID:
+  case AIE2P::EXPVEC64RegClassID:
+  case AIE2P::EXPVEC64_with_sub_hi_exp_in_eEheRegClassID:
+  case AIE2P::EXPVEC64_with_sub_hi_exp_in_eEhoRegClassID:
```
Are these 2 register classes needed? `EXPVEC64_with_sub_hi_exp_in_eEheRegClassID` and `EXPVEC64_with_sub_hi_exp_in_eEhoRegClassID`?

niwinanto replied:
Yes, it was giving crashes while instruction selection.

khallouh commented on llvm/include/llvm/IR/IntrinsicsAIE2P.td:
```diff
@@ -367,6 +367,40 @@ def int_aie2p_I512_I512_ACC2048_negmul_conf : ClangBuiltin<"__builtin_aie2p_I512
+// BFP16 MAC MUL
+class AIE2PMUL_BFP16V64BFP16
+      : Intrinsic<[llvm_v64i32_ty], [llvm_v64i8_ty, llvm_v8i8_ty, llvm_v64i8_ty, llvm_v8i8_ty, llvm_i32_ty],
+					 [IntrNoMem]>;
```
These intrinsics read control registers. To model this correctly, we set the following attributes `[IntrReadMem, IntrInaccessibleMemOnly]` Could you update?

khallouh commented on llvm/lib/Target/AIE/aie2p/AIE2PInstructionSelector.cpp:
```diff
@@ -2973,6 +2984,88 @@ AIE2PInstructionSelector::getCombinedOpcodeSRSUPS(
 }
 
+bool AIE2PInstructionSelector ::selectVMAC_BFP(MachineInstr &I,
+                                               MachineRegisterInfo &MRI) {
```
We can have patterns for all these intrinsics. I tried this one and it works after defining the 2 higher composed sub-register indices. See below:
```
def : PatInaccessibleMem<(int_aie2p_BFP576_BFP1152_ACC2048_mul_conf VEC512:$s1, EXPVEC64:$s2, VEC512:$s3, VEC512:$s4, EXPVEC64:$s5, EXPVEC64:$s6, eR:$acc),
           (VMUL_f_vmul_bfp_vmul_bfp_core_EX_EY 
            (REG_SEQUENCE mEXv, VEC512:$s1, sub_bfp16_x, EXPVEC64:$s2, sub_bfp16_e), 
            (REG_SEQUENCE eEY, 
              VEC512:$s3, sub_bfp16_x, EXPVEC64:$s5, sub_bfp16_e, VEC512:$s4, sub_bfp576_hi_bfp16_x, EXPVEC64:$s6, sub_bfp576_hi_bfp16_e), 
            eR:$acc)>;
```
In `AIE2PRegisterInfo.td`:
```
def sub_bfp576_hi_bfp16_e : ComposedSubRegIndex<sub_bfp576_hi, sub_bfp16_e>;
def sub_bfp576_hi_bfp16_x : ComposedSubRegIndex<sub_bfp576_hi, sub_bfp16_x>;
```

niwinanto replied:
@khallouh This is really nice. I could not make it working with `REG_SEQUENCE`. Thanks for spending time on it.

khallouh reviewed (commented):
Could you add the intrinsics to the `isAccIntrinsic` in RegBankSelect? And to test this store the result of the MAC operations instead of copying it into an accumulator register.
The rest looks good, thanks for addressing the comments.

niwinanto commented:
> Could you add the intrinsics to the `isAccIntrinsic` in RegBankSelect? And to test this store the result of the MAC operations instead of copying it into an accumulator register. The rest looks good, thanks for addressing the comments.

Done. For the register bank tests instead of copying to accumulator register, returning the virtual register. This will do the trick.

khallouh reviewed (approved):
LGTM!
